### DISCUSSION TITLE
How exactly do global registers work?

### DISCUSSION THREAD

**Discussion by PistonMiner** [Vector35/binaryninja-api#1661] [Architectures] [answered]
My understanding of the "global registers" feature (`Architecture.global_regs`) is that they are registers which may be substituted with a fixed value in IL.

How is this fixed value discovered, and more to the point, how can I convince Binary Ninja that a global register has a certain value when it fails to automatically discover it and instead leaves it as the register with undetermined value in IL?

**Comment by rssor** [comment, marked as answer]
It works in concert with the global pointer register (if any) specified in the calling convention. At present, a CC can only list a single global pointer register, and that should probably be expanded. Essentially, once a write to global pointer register is found in a function, it can be copied up as the global pointer value for the binaryview, and then substituted in for reads from the global pointer register when analyzing functions.

As some ABIs include multiple global registers with fixed (or fixed within certain regions) values, it's going to be expanded further at some point.

**Reply by PistonMiner** [reply]
Thanks for the response. Am I understanding correctly that at the moment it is not possible to have more than one fixed register in any one function, since a calling convention has at most one global pointer register? And that `global_regs` merely exists as a list of all the possible global pointer registers to discover fixed values for?

My motivation here is that I have an architecture/ABI that has two such fixed registers in every function and I'm looking for a way for Binary Ninja to substitute both of them for their fixed value, and not just one of them as it does now; however it sounds like that may not be possible right now unfortunately.

**Reply by rssor** [reply]
You are correct in that a BinaryView right now will have one global register value.

For calling conventions there’s another option though: they have a callback for providing the initial value of registers, which is how we handle `$t9` in the case of MIPS. You’d be able to do this and supply fixed values for any number of registers that you want.
